### PR TITLE
Add a plugin to remove trackers from results URLs

### PR DESCRIPTION
**Pull request by Cqoicebordel**
Add a tiny plugin that remove the trackers arguments (like 'utm_*') in URLs returned after a search.

I was not able to test it with a real life test, because I couldn't find an example where a search engine would return such URLs.

First draft of #316

### CODE REVIEW

File: searx/plugins/tracker_url_remover.py [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+You should have received a copy of the GNU Affero General Public License
+along with searx. If not, see < http://www.gnu.org/licenses/ >.
+
+(C) 2015 by Adam Tauber, <asciimoo@gmail.com>
+'''
+
+from flask.ext.babel import gettext
+import re
+
+re1 = re.compile(r'utm_[^&]+&?')
```
Comment (asciimoo) [inline searx/plugins/tracker_url_remover.py]: this could be a list/tuple, which contains the regexps.

File: searx/plugins/tracker_url_remover.py [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+re2 = re.compile(r'(wkey|wemail)[^&]+&?')
```
Comment (asciimoo) [inline searx/plugins/tracker_url_remover.py]: ^^^ replaces e.g. `http://wkey.com/` to `http://`
maybe checking the query params from `result['parsed_url']` would be a safer solution

File: searx/plugins/tracker_url_remover.py [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+re3 = re.compile(r'&$')
+re4 = re.compile(r'^\?$')
```
Comment (asciimoo) [inline searx/plugins/tracker_url_remover.py]: is the above regex necessary? it matches only the `?` string

**Comment by Cqoicebordel** [comment]
Hum... I agree, it needs retooling.

But :
- result['parsed_url'] has always empty arguments
- I found those regex in a JS to remove tracker, I forgot they only applied to the arguments parts. I'll retool them.
- That explain the '?' string : after removing all arguments, if there is only a '?', we can also remove it. The order of the subs is important.

Agreed also on the list for the regexes.

(It pisses me off when I produce bad code like that)

**Comment by Cqoicebordel** [comment]
@asciimoo ok, tell me if this is better.

File: searx/plugins/tracker_url_remover.py [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+           re.compile(r'(wkey|wemail)[^&]+&?'),
+           re.compile(r'&$')}
+
+name = gettext('Tracker URL remover')
+description = gettext('Remove trackers arguments from the returned URL')
+default_on = True
+
+
+def on_result(request, ctx):
+    splited_url = ctx['result']['url'].split('?')
```
Comment (dalf) [inline searx/plugins/tracker_url_remover.py]: Perhaps, just in case `.split('?', 1)` or use`result['result']['parsed_url'].query`

**Comment by Cqoicebordel** [comment]
For the `split('?', 1)`, I thought about it, but it's more useful to test the lenght after, and doing nothing if the length is not 2. It eliminates the "no query" and the "hmmm there is something bad in this URL" type or results.
As for `result['result']['parsed_url'].query`, when I tested it, it always had empty arguments.

So I think I used the most usable way of doing things. But, as always, I may be wrong.

**Comment by asciimoo** [comment]
that's strange, `result['result']['parsed_url'].query` shouldn't be empty. I've tested with `https://aa.bb/cc?dd=ee#ff` and it contained the `dd=ee` string.

**Comment by Cqoicebordel** [comment]
How can you try with a link you provide ?

I'll test again. Maybe I messed up, but I don't think I did.

Anyway, and off topic, isn't it redundant to have the url and the parsed_url in the same var ? It could be tricky too, if you change one but not the other.

**Comment by Cqoicebordel** [comment]
Yeah. I messed up.

I looked at params, and not at query...

Is it good this time ?

**Comment by asciimoo** [comment]
yap, perfect, thanks!
Storing the full and the parsed url is a bit redundant, i agree, but can be handy sometimes (e.g. result merge)

**Comment by Cqoicebordel** [comment]
Ah finally ! :)

For the redundancy, for me, the issue is not so much that it's redundant, it's that inevitably, in the future, a bug will arise from that, because one value would have been changed but not the other.
And yes, it's handy, but it's pretty easy to infer one from the other, at any time : it's just a urlunparse/urlparse away...

**Comment by dalf** [comment]
One way, it's to have an Class like in this project :  https://github.com/gruns/furl/

**Comment by asciimoo** [comment]
ah, yes, furl would be an elegant solution!

**Pull request merged**